### ISSUE TITLE
Predicting for structures in parallel

### ISSUE THREAD

**Issue by vsumaria** [materialsvirtuallab/maml#297]
Is there a way to generate descriptors for the "validation set" data in parallel to predict the energy/ forces/ stress ? I am currently using EnergyForceStress, but have to loop over each structures individually to do these predictions.

**Comment by JiQi535** [comment]
When validation set is mentioned, I assume the predicted energies and forces of structures this set will be compared with the known energies and forces of this set. For such evaluation, e.g., MAE in energies and forces, you can try the `.evaluate() `method for MLIAPs in MAML. You can refer to the "**Predict and evaluate the energies and forces of training data**" part of [our notebook](https://github.com/materialsvirtuallab/maml/blob/master/notebooks/pes/mtp/example.ipynb).

And to answer your question of EnergyForceStress, if I understand your question correctly, it is not necessary to loop over each structure by yourself. The EnergyForceStress takes a structure list as input to calculate, as shown in the "**Energy, force, stress prediction**" part of [our notebook](https://github.com/materialsvirtuallab/maml/blob/master/notebooks/pes/mtp/example.ipynb). But I am aware that the EnergyForceStress basically loops over the structure lists to calculate, so parallelization might be helpful if the validation set is large and the validation process repeats frequently. In that case, python packages like `multiprocess` should be easy to implement and to conduct the `EnergyForceStress.calculate()` for a list of structures with customized number of processes in parallel.

Hope that helps!

**Comment by JiQi535** [comment]
> Is there a way to generate descriptors for the "validation set" data in parallel to predict the energy/ forces/ stress ? I am currently using EnergyForceStress, but have to loop over each structures individually to do these predictions.

When validation set is mentioned, I assume the predicted energies and forces of structures this set will be compared with the known energies and forces of this set. For such evaluation, e.g., MAE in energies and forces, you can try the `.evaluate() `method for MLIAPs in MAML. You can refer to the "**Predict and evaluate the energies and forces of training data**" part of [our notebook](https://github.com/materialsvirtuallab/maml/blob/master/notebooks/pes/mtp/example.ipynb).

And to answer your question of EnergyForceStress, if I understand your question correctly, it is not necessary to loop over each structure by yourself. The EnergyForceStress takes a structure list as input to calculate, as shown in the "**Energy, force, stress prediction**" part of [our notebook](https://github.com/materialsvirtuallab/maml/blob/master/notebooks/pes/mtp/example.ipynb). But I am aware that the EnergyForceStress basically loops over the structure lists to calculate, so parallelization might be helpful if the validation set is large and the validation process repeats frequently. In that case, python packages like `multiprocess` should be easy to implement and to conduct the `EnergyForceStress.calculate()` for a list of structures with customized number of processes in parallel.

Hope that helps!

**Comment by vsumaria** [comment]
Sorry about this. I just went though the notebook to see the .evaluate() function. Does this function also loop over structures one by one ? 

And I agree, I was thinking of implementing the same with multiprocess.

**Comment by JiQi535** [comment]
> Sorry about this. I just went though the notebook to see the .evaluate() function. Does this function also loop over structures one by one ?
> 
> And I agree, I was thinking of implementing the same with multiprocess.

I just checked and found that the `.evaluate() `of snap potentials actually can process structures in parallel. And I think implementing parallelization with multiprocess is doable.

**Comment by vsumaria** [comment]
I just wrote a script for parallel evaluation of trained SNAP potential. I am Posting it here if it is helpful for anyone else. (I read structures from an ASE traj file and am plotting a parity plot for energy/atom and forces(x,y,z). 

`from maml.apps.pes import SNAPotential
from maml.apps.pes import EnergyForceStress
from pymatgen.io.ase import AseAtomsAdaptor
from ase.io.trajectory import Trajectory
import matplotlib.pyplot as plt
import numpy as np
from ase.io import *
from joblib import Parallel, delayed

qsnap_loaded = SNAPotential.from_config(param_file='../SNAPotential.snapparam', coeff_file='../SNAPotential.snapcoeff')
efs_calculator = EnergyForceStress(ff_settings=qsnap_loaded)

def calc_nn(struc):
    energy, forces, stresses = efs_calculator.calculate([struc])[0]
    return energy, forces

traj = read('../TRAIN12.traj@0:64')
ase_adap = AseAtomsAdaptor()

train_energies = [atoms.get_potential_energy() for i,atoms in enumerate(traj)]
train_forces =  [atoms.get_forces() for i,atoms in enumerate(traj)]
train_structures = [ase_adap.get_structure(atoms) for i,atoms in enumerate(traj)]

def calc(struc):
    energy, forces, stresses = efs_calculator.calculate([struc])[0]
    return energy, forces

res = Parallel(n_jobs=64)(delayed(calc_nn)(atoms) for atoms in train_structures)
e_nn = np.array([item[0] for item in res])
f_nn = [item[1] for item in res]

#force parity plot
f_dft = [atoms.get_forces() for atoms in traj]
e_dft = np.array([atoms.get_potential_energy() for atoms in traj])
f_nn = np.array(np.concatenate(f_nn).ravel().tolist())
f_dft = np.array(np.concatenate(f_dft).ravel().tolist())
f_all = np.concatenate((f_dft,f_nn))
f_min = np.min(np.min(f_all))
f_max = np.max(np.max(f_all))
xx = [f_min,f_max]
x_text = np.mean(xx)-0.05*(f_max-f_min)
y_text = np.mean(xx)-0.3*(f_max-f_min)
e = abs(f_dft-f_nn)
rmse = np.sqrt(np.mean(e**2))
plt.plot(f_nn, f_dft,'o')
plt.plot(xx,xx,'k--')
plt.text(x_text,y_text,'RMSE = %.02f eV/$\AA$'% (rmse))
plt.savefig('parity_f.png')

#energy parity
plt.figure()
natoms = [len(atoms) for atoms in traj]
e_dft = e_dft/natoms
e_nn = e_nn/natoms
e_all = np.concatenate((e_dft,e_nn))
e_min = np.min(np.min(e_all))
e_max = np.max(np.max(e_all))
xx = [e_min,e_max]
x_text = np.mean(xx)-0.05*(e_max-e_min)
y_text = np.mean(xx)-0.3*(e_max-e_min)
e = abs(e_dft-e_nn)
rmse = np.sqrt(np.mean(e**2))*1000
plt.plot(e_nn, e_dft,'o')
plt.plot(xx,xx,'k--')
plt.text(x_text,y_text,'RMSE = %.02f meV/atom'% (rmse))
plt.savefig('parity_e.png')

plt.show()`
